"Unnecessary dict call" false positives

The dict function has 3 forms: `dict(**kwarg)`, `dict(mapping, **kwarg)`, and `dict(iterable, **kwarg)`.  I'm not sure if it makes sense to apply this rule to every form.

Here's one example use case:
```py
def foo(data):
    defaults = {'a': 1, 'b': 2}
    return dict(defaults, **data)
```

This example could be rewritten to use literals, but it would be less succinct and (I assume, but haven't confirmed) less performant.

I tried a "manual" way:
1. Create empty python3.6.1 virtualenv.
`$ mkvirtualenv -p python3 flake`
2. Install flake8-comprehensions.
`pip install flake8-comprehensions`
3. Verify result.
```
$ pip list --format columns
Package               Version
--------------------- -------
appdirs               1.4.3  
flake8                3.3.0  
flake8-comprehensions 1.4.0  
mccabe                0.6.1  
packaging             16.8   
pip                   9.0.1  
pycodestyle           2.3.1  
pyflakes              1.5.0  
pyparsing             2.2.0  
setuptools            35.0.2 
six                   1.10.0 
wheel                 0.29.0 
```
4. Create an example.py file.
```
echo "a = {'foo': 'bar'}
> dict(baz='bar', **a)" > example.py
```
5. Run flake8.
```
$ flake8 example.py 
example.py:2:1: C408 Unnecessary dict call - rewrite as a literal.
```

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -7,6 +7,8 @@ Pending Release
 
 .. Insert new release notes below this line
 
+* Fix false positives in ``C408`` for calls using ``*args`` or ``**kwargs``.
+
 1.4.0 (2017-05-14)
 ------------------
 

--- a/flake8_comprehensions.py
+++ b/flake8_comprehensions.py
@@ -141,6 +141,8 @@ class ComprehensionChecker(object):
 
                 elif (
                     n_args == 0 and
+                    not has_star_args(node) and
+                    not has_keyword_args(node) and
                     node.func.id in ('tuple', 'list', 'dict')
                 ):
                     yield (
@@ -149,3 +151,17 @@ class ComprehensionChecker(object):
                         self.messages['C408'].format(type=node.func.id),
                         type(self),
                     )
+
+
+def has_star_args(call_node):
+    return (
+        any(isinstance(a, ast.Starred) for a in call_node.args) or  # Python 3
+        getattr(call_node, 'starargs', None) is not None  # Python 2
+    )
+
+
+def has_keyword_args(call_node):
+    return (
+        any(k.arg is None for k in call_node.keywords) or  # Python 3
+        getattr(call_node, 'kwargs', None) is not None  # Python 2
+    )

--- a/tests/test_it.py
+++ b/tests/test_it.py
@@ -350,6 +350,22 @@ def test_C408_pass_4():
     assert errors == []
 
 
+def test_C408_pass_5():
+    errors = run_flake8('''\
+        foo = {}
+        dict(bar=1, **foo)
+    ''')
+    assert errors == []
+
+
+def test_C408_pass_6():
+    errors = run_flake8('''\
+        foo = [1, 2]
+        list(*foo)
+    ''')
+    assert errors == []
+
+
 def test_C408_fail_1():
     errors = run_flake8('tuple()')
     assert errors == [

